quinn/src/lib.rs

```diff
@@ -1074,13 +1074,15 @@ pub trait Read {
     ///
     /// Returns a segment of data and their offset in the stream. Segments may be received in any order and may overlap.
     ///
-    /// Using this function reduces latency improves throughput by avoiding head-of-line blocking within the stream, and
-    /// reduces computational overhead by allowing data to be passed on without any intermediate buffering. Prefer it
-    /// whenever possible.
+    /// Using this function reduces latency and improves throughput by avoiding head-of-line blocking within the stream,
```

Comment:
You didn't change this much content-wise, but I'll offer a suggestion anyway: I think this comment doesn't do a very good job of highlighting the trade-offs involved. QUIC-as-specified is still largely geared towards in-order delivery (DATAGRAM discussions notwithstanding), and while it can be a viable choice to opt into unordered delivery for some applications, "Prefer it whenever possible." while probably correct feels a little biased to me.

Reply:
I think it's important to distinguish the delivery of data in the order it was sent from the delivery of information in a manner that preserves the data's position in its stream. QUIC is geared towards doing the latter, but I don't think the former is specifically favored or even all that useful.

Reply:
Hmm, I don't really agree. In a lot of places, QUIC is being pitched as a TCP replacement, which I think implies reliable in-order delivery to the application layer -- which is also the simplest to use, which is also an important consideration.

All of which to say, I think reliable in-order delivery should be our default public API in a sense, with unordered (and, in the future, unreliable) alternatives for those who really needed.

Reply:
On consideration, the current advice is certainly too strong in light of ordered messaging applications, which are common. I'll go ahead and weaken the language. There are certainly lots of traditional TCP applications (e.g. bulk data transfer) where unordered reads make much more sense, but they're indeed not universal.